# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/html/HTMLImageElement.h
# Commit: 2b4d9516b882
# Full Hash: 2b4d9516b882be621bcec24dc7310f32243c9549
# Author: Boris Chiou <boris.chiou@gmail.com>
# Date: 2022-08-23 02:46:57
# Regressor Bug: 1694741
# File Overlap Count: 1
# Description:
#   Bug 1694741 - Part 5: Update responsive image selector synchronously. r=emilio,edgar
#   
#   In order to make sure we are using the correct source element when
#   flushing the style, it's necessary to update the responsive image selector
#   synchronously (i.e. full selection for source element and re-resolve the
# ==============================================================================

diff -r a8f10a064b69 -r 2b4d9516b882 dom/html/HTMLImageElement.h
--- a/dom/html/HTMLImageElement.h	Mon Aug 22 20:18:38 2022 +0000
+++ b/dom/html/HTMLImageElement.h	Mon Aug 22 20:18:39 2022 +0000
@@ -273,13 +273,15 @@
  protected:
   virtual ~HTMLImageElement();
 
-  // Queues a task to run LoadSelectedImage pending stable state.
+  // Update the responsive source synchronously and queues a task to run
+  // LoadSelectedImage pending stable state.
   //
   // Pending Bug 1076583 this is only used by the responsive image
   // algorithm (InResponsiveMode()) -- synchronous actions when just
   // using img.src will bypass this, and update source and kick off
   // image load synchronously.
-  void QueueImageLoadTask(bool aAlwaysLoad);
+  void UpdateSourceSyncAndQueueImageTask(
+      bool aAlwaysLoad, const HTMLSourceElement* aSkippedSource = nullptr);
 
   // True if we have a srcset attribute or a <picture> parent, regardless of if
   // any valid responsive sources were parsed from either.
@@ -292,8 +294,8 @@
   // True if the given URL equals the last URL that was loaded by this element.
   bool SelectedSourceMatchesLast(nsIURI* aSelectedSource);
 
-  // Resolve and load the current mResponsiveSelector (responsive mode) or src
-  // attr image.
+  // Load the current mResponsiveSelector (responsive mode) or src attr image.
+  // Note: This doesn't run the full selection for the responsive selector.
   nsresult LoadSelectedImage(bool aForce, bool aNotify, bool aAlwaysLoad);
 
   // True if this string represents a type we would support on <source type>
@@ -308,9 +310,9 @@
   // we don't actually care which changed or to what
   void PictureSourceMediaOrTypeChanged(nsIContent* aSourceNode, bool aNotify);
 
-  void PictureSourceAdded(nsIContent* aSourceNode);
+  void PictureSourceAdded(HTMLSourceElement* aSourceNode = nullptr);
   // This should be called prior to the unbind, such that nextsibling works
-  void PictureSourceRemoved(nsIContent* aSourceNode);
+  void PictureSourceRemoved(HTMLSourceElement* aSourceNode = nullptr);
 
   // Re-evaluates all source nodes (picture <source>,<img>) and finds
   // the best source set for mResponsiveSelector. If a better source
@@ -324,8 +326,13 @@
   // parameters as appropriate before calling (or null it out to force
   // recreation)
   //
+  // if |aSkippedSource| is non-null, we will skip it when running the
+  // algorithm. This is used when we need to update the source when we are
+  // removing the source element.
+  //
   // Returns true if the source has changed, and false otherwise.
-  bool UpdateResponsiveSource();
+  bool UpdateResponsiveSource(
+      const HTMLSourceElement* aSkippedSource = nullptr);
 
   // Given a <source> node that is a previous sibling *or* ourselves, try to
   // create a ResponsiveSelector.